src/Discovery/Archive/RPM.hs

```diff
@@ -14,30 +16,59 @@ import qualified Data.ByteString as BS
 import qualified Data.ByteString.Char8 as BS8
 import qualified Data.ByteString.Lazy as BL
 import qualified Data.CPIO as CPIO
+import qualified Data.Conduit.Lzma as Lzma
+import qualified Data.Conduit.Zlib as Zlib
+import Data.Foldable (asum)
 import Path
-import Prelude
 import qualified Path.IO as PIO
+import Prelude
 
 extractRpm :: Has (Lift IO) sig m => Path Abs Dir -> Path Abs File -> m ()
 extractRpm dir rpmFile = do
   res <-
-    sendIO . runResourceT . runExceptT . runConduit $
-      readRPMEntries rpmFile .| filterC (not . CPIO.isEntryDirectory) .| sinkDir
+    sendIO . runResourceT . runExceptT . runConduit $ do
+      sourceFileBS (toFilePath rpmFile) .| RPM.parseRPMC .| awaitForever (extractEntries dir)
```

Comment:
Is this function in any danger of hanging with the use of `awaitForever`? I have seen us come across unsupported file formats when doing RPM extraction in the RPM fetcher that caused extraction to hang. I'm not aware of a better option here, but is this a valid concern? If so, it may be an acceptable failure mode for now.

Reply:
`awaitForever` is a bit of a misnomer: it just repeatedly tries to `await` input from the upstream pipe, passing it to the provided function `(extractEntries dir)` until no more input is available.

`await` in conduit doesn't actually make a thread wait or anything; everything in a conduit happens in the same thread. `await` is just a way to pass control to the upstream until it can give you back an element, at which point control flow moves back to you